rewrite-maven/src/main/java/org/openrewrite/maven/RemoveDuplicateDependencies.java

```diff
@@ -0,0 +1,124 @@
+/*
+ * Copyright 2022 the original author or authors.
+ * <p>
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ * <p>
+ * https://www.apache.org/licenses/LICENSE-2.0
+ * <p>
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.openrewrite.maven;
+
+import lombok.EqualsAndHashCode;
+import lombok.Value;
+import org.openrewrite.ExecutionContext;
+import org.openrewrite.Recipe;
+import org.openrewrite.TreeVisitor;
+import org.openrewrite.internal.ListUtils;
+import org.openrewrite.marker.SearchResult;
+import org.openrewrite.xml.XPathMatcher;
+import org.openrewrite.xml.tree.Content;
+import org.openrewrite.xml.tree.Xml;
+
+import java.time.Duration;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+import java.util.function.UnaryOperator;
+
+@Value
+@EqualsAndHashCode(callSuper = true)
+public class RemoveDuplicateDependencies extends Recipe {
+
+    @Override
+    public String getDisplayName() {
+        return "Remove duplicate Maven dependencies";
+    }
+
+    @Override
+    public String getDescription() {
+        return "Removes duplicated dependencies in the `<dependencies>` and `<dependencyManagement>` sections of the `pom.xml`.";
+    }
+
+    @Override
+    public Duration getEstimatedEffortPerOccurrence() {
+        return Duration.ofMinutes(2);
+    }
+
+    @Override
+    public TreeVisitor<?, ExecutionContext> getVisitor() {
+        return new RemoveDuplicateDependenciesVisitor();
+    }
+
+    @Override
+    protected MavenIsoVisitor<ExecutionContext> getApplicableTest() {
+        return new MavenIsoVisitor<ExecutionContext>() {
+            @Override
+            public Xml.Document visitDocument(Xml.Document document, ExecutionContext executionContext) {
+                Xml.Tag root = document.getRoot();
+                if (root.getChild("dependencies").isPresent() || root.getChild("dependencyManagement").isPresent()) {
+                    return SearchResult.found(document);
+                }
+                return document;
+            }
+        };
+    }
+
+    private static class RemoveDuplicateDependenciesVisitor extends MavenIsoVisitor<ExecutionContext> {
+        private static final XPathMatcher DEPENDENCIES_MATCHER = new XPathMatcher("/project/dependencies");
+        private static final XPathMatcher MANAGED_DEPENDENCIES_MATCHER = new XPathMatcher("/project/dependencyManagement/dependencies");
+
+        @Override
+        public Xml.Tag visitTag(Xml.Tag tag, ExecutionContext ctx) {
```

Comment:
The case you mention regarding first declaring the dependency with the test scope is interesting. I will check how Maven deals with that.

Regarding the other use case I am fairly certain that Maven has a depth-first and first occurrence wins strategy. With that in mind the version of the duplicates is actually irrelevant. But truth be told, I am not 100% sure regarding this.

I would actually have expected this rule to also apply to the scopes use case, which would mean that if first declared as test-scoped the dependency would not be visible to the runtime classpath. But I will verify this.

Reply:
The maven resolution model is created using the same rules that Maven uses. Note, that if dependencies are declared in a parent pom.xml, those are appended to the list of dependencies in the current pom's effective model. This, plus the "encountered first" rule, dependency conflict resolution, ranges, property resolution, etc are all handled when deriving the maven model.

Now that you mentioned it, you may be right about having the same dependency in the same pom (with different scopes).

Reply:
I also forgot to mention that after a structural change has been made to the pom's AST, we will also want to call "maybeUpdateModel()" which will re-run maven resolution and update the semantic model.

Reply:
Personally I think I would still be in favor of a "syntactic" duplication check, which is limited to the current Maven module only. I would for instance not be happy if the recipe would remove a dependency just because it appeared as a transitive dependency of some other previous dependency in my module.

But possibly the recipe should offer both alternatives, what do you think?

Reply:
You've convinced me. I should use the resolved dependency model instead. I will look into it.

Reply:
I don't quite understand what exactly you meant by _and then consult that to determine if the resolved dependency originated from the current tag_. I didn't find any mapping between the two.

Anyway, I added a custom `findDependency(Xml.Tag)` method, because I also want to match the `type` and not match the `version`.